streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -1063,4 +1072,49 @@ object ZSink extends ZSinkPlatformSpecific {
       def stepPure(state: State, a: Any): Step[State, Nothing] = Step.done(state, Chunk.empty)
       def extractPure(state: State): Either[Nothing, B]        = Right(b)
     }
+
+  /**
+   * Creates a sink which throttles input elements of type A according to the given bandwidth parameters
+   * using the token bucket algorithm. The weight of each element is determined by the `costFn` effectful
+   * function. Stream elements are mapped to `Option[A]`, and `None` denotes that a given element has
+   * been throttled.
+   */
+  final def throttle[R, E, A](units: Long, duration: Duration)(
```

Comment:
Let's call this `throttleEnforceM` and add a version `throttleEnforce` with `costFn: A => Long`

Reply:
Done.